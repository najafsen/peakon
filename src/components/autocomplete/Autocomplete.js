import React, { useState, useEffect, useCallback, useRef } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import './Autocomplete.scss';
import { createSearchRegex } from '../../lib/regex.lib';
import { AutocompleteItem } from './AutocompleteItem';

export const VISIBLE_ITEMS = 2;

export const KEY_CODES = {
    TAB: 9,
    ENTER: 13,
    ESCAPE: 27,
    ARROW_UP: 38,
    ARROW_DOWN: 40
}

export const Autocomplete = React.memo(({ items, onSelect, placeholder = 'Search...' }) => {
    const [filteredItems, setFilteredItems] = useState(items);
    const [query, setQuery] = useState('');
    const [isOpen, setIsOpen] = useState(false);
    const [activeItem, setActiveItem] = useState(-1);
    const inputRef = useRef(null);
    const arrowRef = useRef(null);
    const menuRef = useRef(null);

    const selectItem = useCallback((item) => {
        setQuery(item.name);
        onSelect(item);
        inputRef.current.focus();
    }, [onSelect]);

    const handleDocumentClick = useCallback((e) => {
        if (
            inputRef.current &&
            arrowRef.current &&
            !inputRef.current.contains(e.target) &&
            !arrowRef.current.contains(e.target)
        ) {
            setIsOpen(false);
            setActiveItem(-1);
        }
    }, []);

    const scrollToItem = useCallback((itemIndex, direction) => {
        if (!menuRef.current) {
            return;
        }

        const scrollTop = menuRef.current.scrollTop;
        const scrollHeight = menuRef.current.scrollHeight;
        const itemHeight = scrollHeight / filteredItems.length;
        const lastVisibleItemScrollTop = scrollTop + itemHeight * (VISIBLE_ITEMS - 1);
        const topIndex = direction === 'up' ? itemIndex : Math.max(0, itemIndex - VISIBLE_ITEMS + 1);
        const newTopToCompare = itemIndex * itemHeight;
        const newTopToScroll = topIndex * itemHeight;

        if (newTopToCompare >= scrollTop && newTopToCompare <= lastVisibleItemScrollTop) {
            return;
        }

        menuRef.current.scroll({
            top: newTopToScroll,
            behavior: 'smooth'
        });
    }, [filteredItems.length]);

    const handleInputKeyDown = useCallback((e) => {
        let newActiveItem = activeItem;

        switch (e.keyCode) {
            case KEY_CODES.ENTER: {
                if (!isOpen) {
                    break;
                }
                setActiveItem(-1);
                setIsOpen(false);
                selectItem(filteredItems[activeItem]);
                break;
            }

            case KEY_CODES.TAB:
            case KEY_CODES.ESCAPE: {
                setActiveItem(-1);
                setIsOpen(false);
                break;
            }

            case KEY_CODES.ARROW_UP: {
                if (!isOpen) {
                    setIsOpen(true);
                }
                newActiveItem = activeItem <= 0 ? filteredItems.length - 1 : activeItem - 1;
                setActiveItem(newActiveItem);
                setTimeout(() => scrollToItem(newActiveItem, 'up'));
                break;
            }

            case KEY_CODES.ARROW_DOWN: {
                if (!isOpen) {
                    setIsOpen(true);
                }
                newActiveItem = activeItem === filteredItems.length - 1 ? 0 : activeItem + 1;
                setActiveItem(newActiveItem);
                setTimeout(() => scrollToItem(newActiveItem, 'down'));
                break;
            }

            default: break;
        }
    }, [activeItem, isOpen, selectItem, filteredItems, scrollToItem]);

    const handleInputChange = useCallback(e => {
        if (!isOpen) {
            setIsOpen(true);
        }

        setQuery(e.target.value);
    }, [isOpen]);

    const handleItemMouseOver = useCallback((item) => {
        setActiveItem(filteredItems.indexOf(item));
    }, [filteredItems]);

    useEffect(() => {
        document.addEventListener('click', handleDocumentClick);
    }, [handleDocumentClick]);

    useEffect(() => {
        const filteredItems = items.filter(item => createSearchRegex(query, 'i').test(item.name));
        setFilteredItems(filteredItems);
    }, [items, query]);

    return (
        <div className="autocomplete-component">

            <input
                data-testid="input"
                ref={inputRef}
                placeholder={placeholder}
                value={query}
                onClick={() => setIsOpen(true)}
                onChange={handleInputChange}
                onKeyDown={handleInputKeyDown}
            />

            <i
                data-testid="arrow"
                ref={arrowRef}
                className={cx('arrow-down', { active: isOpen })}
                onClick={() => setIsOpen(!isOpen)}
            />

            {
                (isOpen && filteredItems.length) ? (
                    <ul
                        data-testid="items"
                        ref={menuRef}
                        className="autocomplete-items"
                    >
                        {
                            filteredItems.map((item, index) => (
                                <AutocompleteItem
                                    key={item.id}
                                    item={item}
                                    isActive={activeItem === index}
                                    onSelect={selectItem}
                                    onMouseOver={handleItemMouseOver}
                                />
                            ))
                        }
                    </ul>
                ) : null
            }

        </div>
    )
});

Autocomplete.propTypes = {
    items: PropTypes.arrayOf(PropTypes.object).isRequired,
    onSelect: PropTypes.func,
    placeholder: PropTypes.string
}