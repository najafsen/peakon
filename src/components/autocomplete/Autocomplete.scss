$base-size: 1rem;
$visible-items: 2;
$border-radius: 0.375 * $base-size;
$border-color: #d2d3cf;
$bg-color: #ffffff;

$input-color: #3F3F3F;

$item-height: 3 * $base-size;
$item-border-color: #e9e9e9;
$item-highlight-color: #f5f5f5;
$item-secondary-color: #939a9d;
$item-secondary-font-size: 0.8 * $base-size;

$input-border-color: #c1c2c4;
$input-focus-shadow-color: #66bf89;
$input-focus-border-color: #85c39d;

$arrow-down-color: $border-color;
$arrow-down-active-color: #656565;

@function get-max-height() {
  @return calc(#{($item-height + (2 * $base-size)) * $visible-items} + #{$visible-items + 1}px);
}

.autocomplete-component {
  position: relative;

  input {
    display: block;
    width: 100%;
    padding: 0.375 * $base-size 0.75 * $base-size;
    font-size: $base-size;
    line-height: 1.8 * $base-size;
    color: $input-color;
    background-color: $bg-color;
    border: 1px solid $input-border-color;
    border-radius: .25 * $base-size;
    box-sizing: border-box;
    transition: border-color 0.25s ease-in-out, outline-color 0.25s ease-in-out, box-shadow 0.25s ease-in-out;

    &:focus {
      outline-color: $input-focus-border-color;
      border-color: $input-focus-border-color;
      box-shadow: 0 0 2px 1px $input-focus-shadow-color;
    }
  }

  .arrow-down {
    position: absolute;
    display: inline-block;
    right: $base-size;
    top: $base-size;
    border: solid $input-border-color;
    border-width: 0 3px 3px 0;
    padding: 3px;
    transform: rotate(45deg);
    transition: transform 0.3s linear;
    cursor: pointer;

    &.active {
      border-color: $arrow-down-active-color;
      transform: rotate(585deg);
    }
  }

  .autocomplete-items {
    list-style-type: none;
    position: absolute;
    padding: 0;
    margin-top: $base-size / 2;
    border: 1px solid $border-color;
    border-radius: $border-radius;
    background-color: $bg-color;
    width: 100%;
    box-sizing: border-box;
    max-height: get-max-height();
    overflow: auto;

    .autocomplete-item {
      display: flex;
      padding: 0 $base-size;
      background-color: $bg-color;
      cursor: pointer;

      &.active {
        background-color: $item-highlight-color;
      }

      &:last-child {
        .item-content {
          border: none;
        }
      }

      .item-content {
        display: flex;
        flex: 1;
        padding: $base-size 0;
        height: $item-height;
        border-bottom: 1px solid $item-border-color;

        .avatar {
          flex: 0;
          margin: 0 $base-size;
        }

        .info {
          display: flex;
          flex-direction: column;
          flex: 1;
          justify-content: flex-start;

          .name {
            flex: 0 0 60%;
          }

          .email {
            flex: 0 0 40%;
            color: $item-secondary-color;
            font-size: $item-secondary-font-size;
          }
        }
      }
    }
  }
}
