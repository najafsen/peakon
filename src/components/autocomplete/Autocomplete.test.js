import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { Autocomplete, KEY_CODES, VISIBLE_ITEMS } from './Autocomplete';
import items from '../../fixtures/employeesFlattened';

describe('Autocomplete', () => {
    beforeEach(() => {
        window.HTMLElement.prototype.scroll = jest.fn();
    });

    afterEach(cleanup);

    test('matches snapshot', () => {
        const { asFragment } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('Shows default placeholder', () => {
        const { getByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        expect(getByTestId('input')).toHaveAttribute('placeholder', 'Search...');
    });

    test('Shows given placeholder', () => {
        const { getByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} placeholder="Choose Manager" />);
        expect(getByTestId('input')).toHaveAttribute('placeholder', 'Choose Manager');
    });

    test('matches the snapshot with menu open', () => {
        const { asFragment, getByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.click(getByTestId('input'));
        expect(asFragment()).toMatchSnapshot();
    });

    test('matches the snapshot with query', () => {
        const { asFragment, getByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.change(getByTestId('input'), { target: { value: 'hm' } });

        expect(asFragment()).toMatchSnapshot();
    });

    test('filters items based on given query', () => {
        const { getByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.change(getByTestId('input'), { target: { value: 'hm' } });

        expect(getByTestId('items').childNodes).toHaveLength(2);
    });


    test('shows no suggestion when no item with given query found', () => {
        const { queryByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.change(queryByTestId('input'), { target: { value: 'xxxxxxxx' } });

        expect(queryByTestId('items')).toBe(null);
    });


    test('toggles the menu by clicking on the arrow button', () => {
        const { queryByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.click(queryByTestId('arrow'))
        expect(queryByTestId('items')).toBeInTheDocument();
        fireEvent.click(queryByTestId('arrow'))
        expect(queryByTestId('items')).toBe(null);
    });


    test('closes the menu when clilcked outside the input', () => {
        const { queryByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.click(queryByTestId('arrow'));
        expect(queryByTestId('items')).toBeInTheDocument();
        fireEvent.click(document);
        expect(queryByTestId('items')).toBe(null);
    });


    test('pressing arrow down will open the menu', () => {
        const { queryByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ARROW_DOWN });
        expect(queryByTestId('items')).toBeInTheDocument();
    });


    test('pressing arrow down 3 times will open the menu and set scroll', (done) => {
        jest.spyOn(window.HTMLElement.prototype, 'scrollHeight', 'get').mockReturnValue(728);
        const { queryByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ARROW_DOWN });
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ARROW_DOWN });
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ARROW_DOWN });
        expect(queryByTestId('items')).toBeInTheDocument();
        setTimeout(() => {
            expect(window.HTMLElement.prototype.scroll).toHaveBeenCalledTimes(3 - VISIBLE_ITEMS);
            done();
        })
    });


    test('pressing arrow up will open the menu', () => {
        const { queryByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ARROW_UP });
        expect(queryByTestId('items')).toBeInTheDocument();
    });

    test('pressing arrow up 3 times will open the menu and set scroll', (done) => {
        jest.spyOn(window.HTMLElement.prototype, 'scrollTop', 'get').mockReturnValue(566);
        jest.spyOn(window.HTMLElement.prototype, 'scrollHeight', 'get').mockReturnValue(728);
        const { queryByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ARROW_UP });
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ARROW_UP });
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ARROW_UP });
        expect(queryByTestId('items')).toBeInTheDocument();
        setTimeout(() => {
            expect(window.HTMLElement.prototype.scroll).toHaveBeenCalledTimes(3 - VISIBLE_ITEMS + 1);
            done();
        })
    });

    test('pressing Escape will close the menu', () => {
        const { queryByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.click(queryByTestId('arrow'));
        expect(queryByTestId('items')).toBeInTheDocument();
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ESCAPE });
        expect(queryByTestId('items')).toBe(null);
    });


    test('pressing Tab will close the menu', () => {
        const { queryByTestId } = render(<Autocomplete items={items} onSelect={jest.fn()} />);
        fireEvent.click(queryByTestId('arrow'));
        expect(queryByTestId('items')).toBeInTheDocument();
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.TAB });
        expect(queryByTestId('items')).toBe(null);
    });


    test('navitgate to second item with keyboard and press Enter will fire onSelect callback and close the menu', () => {
        const onSelect = jest.fn();
        const { queryByTestId } = render(<Autocomplete items={items} onSelect={onSelect} />);
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ARROW_DOWN });
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ARROW_DOWN });
        expect(queryByTestId('items')).toBeInTheDocument();
        fireEvent.keyDown(queryByTestId('input'), { keyCode: KEY_CODES.ENTER });
        expect(onSelect).toHaveBeenCalledWith(items[1]);
        expect(queryByTestId('items')).toBe(null);
    });

    // More tests are required to fully test the feature, but because of the lack of time these sample tests are only written.
});