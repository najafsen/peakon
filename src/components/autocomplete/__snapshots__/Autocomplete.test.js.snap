// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="autocomplete-component"
  >
    <input
      data-testid="input"
      placeholder="Search..."
      value=""
    />
    <i
      class="arrow-down"
      data-testid="arrow"
    />
  </div>
</DocumentFragment>
`;

exports[`Autocomplete matches the snapshot with menu open 1`] = `
<DocumentFragment>
  <div
    class="autocomplete-component"
  >
    <input
      data-testid="input"
      placeholder="Search..."
      value=""
    />
    <i
      class="arrow-down active"
      data-testid="arrow"
    />
    <ul
      class="autocomplete-items"
      data-testid="items"
    >
      <li
        class="autocomplete-item"
        data-testid="li"
      >
        <div
          class="item-content"
        >
          <div
            class="avatar"
          >
            <div
              class="avatar-component"
            >
              <div
                class="avatar-alt"
                data-testid="alt"
              >
                HM
              </div>
              <img
                alt="Harriet McKinney"
                class="avatar-image hidden"
                data-testid="image"
              />
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="name"
            >
              Harriet McKinney
            </div>
            <div
              class="email"
            />
          </div>
        </div>
      </li>
      <li
        class="autocomplete-item"
        data-testid="li"
      >
        <div
          class="item-content"
        >
          <div
            class="avatar"
          >
            <div
              class="avatar-component"
            >
              <div
                class="avatar-alt"
                data-testid="alt"
              >
                HB
              </div>
              <img
                alt="Harriet Banks"
                class="avatar-image hidden"
                data-testid="image"
              />
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="name"
            >
              Harriet Banks
            </div>
            <div
              class="email"
            />
          </div>
        </div>
      </li>
      <li
        class="autocomplete-item"
        data-testid="li"
      >
        <div
          class="item-content"
        >
          <div
            class="avatar"
          >
            <div
              class="avatar-component"
            >
              <div
                class="avatar-alt"
                data-testid="alt"
              >
                MS
              </div>
              <img
                alt="Mathilda Summers"
                class="avatar-image hidden"
                data-testid="image"
              />
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="name"
            >
              Mathilda Summers
            </div>
            <div
              class="email"
            >
              alta.maxwell@kinetar.com
            </div>
          </div>
        </div>
      </li>
      <li
        class="autocomplete-item"
        data-testid="li"
      >
        <div
          class="item-content"
        >
          <div
            class="avatar"
          >
            <div
              class="avatar-component"
            >
              <div
                class="avatar-alt"
                data-testid="alt"
              >
                EW
              </div>
              <img
                alt="Eugene Wong"
                class="avatar-image hidden"
                data-testid="image"
              />
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="name"
            >
              Eugene Wong
            </div>
            <div
              class="email"
            >
              harriet.banks@kinetar.com
            </div>
          </div>
        </div>
      </li>
      <li
        class="autocomplete-item"
        data-testid="li"
      >
        <div
          class="item-content"
        >
          <div
            class="avatar"
          >
            <div
              class="avatar-component"
            >
              <div
                class="avatar-alt"
                data-testid="alt"
              >
                NM
              </div>
              <img
                alt="New Manager"
                class="avatar-image hidden"
                data-testid="image"
              />
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="name"
            >
              New Manager
            </div>
            <div
              class="email"
            />
          </div>
        </div>
      </li>
      <li
        class="autocomplete-item"
        data-testid="li"
      >
        <div
          class="item-content"
        >
          <div
            class="avatar"
          >
            <div
              class="avatar-component"
            >
              <div
                class="avatar-alt"
                data-testid="alt"
              >
                MR
              </div>
              <img
                alt="Marguerite Ryan"
                class="avatar-image hidden"
                data-testid="image"
              />
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="name"
            >
              Marguerite Ryan
            </div>
            <div
              class="email"
            />
          </div>
        </div>
      </li>
      <li
        class="autocomplete-item"
        data-testid="li"
      >
        <div
          class="item-content"
        >
          <div
            class="avatar"
          >
            <div
              class="avatar-component"
            >
              <div
                class="avatar-alt"
                data-testid="alt"
              >
                DB
              </div>
              <img
                alt="Donald Butler"
                class="avatar-image hidden"
                data-testid="image"
              />
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="name"
            >
              Donald Butler
            </div>
            <div
              class="email"
            />
          </div>
        </div>
      </li>
      <li
        class="autocomplete-item"
        data-testid="li"
      >
        <div
          class="item-content"
        >
          <div
            class="avatar"
          >
            <div
              class="avatar-component"
            >
              <div
                class="avatar-alt"
                data-testid="alt"
              >
                JC
              </div>
              <img
                alt="Jim Carlson"
                class="avatar-image hidden"
                data-testid="image"
              />
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="name"
            >
              Jim Carlson
            </div>
            <div
              class="email"
            />
          </div>
        </div>
      </li>
      <li
        class="autocomplete-item"
        data-testid="li"
      >
        <div
          class="item-content"
        >
          <div
            class="avatar"
          >
            <div
              class="avatar-component"
            >
              <div
                class="avatar-alt"
                data-testid="alt"
              >
                AM
              </div>
              <img
                alt="Alta Maxwell"
                class="avatar-image hidden"
                data-testid="image"
              />
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="name"
            >
              Alta Maxwell
            </div>
            <div
              class="email"
            >
              eugene.wong@kinetar.com
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Autocomplete matches the snapshot with query 1`] = `
<DocumentFragment>
  <div
    class="autocomplete-component"
  >
    <input
      data-testid="input"
      placeholder="Search..."
      value="hm"
    />
    <i
      class="arrow-down active"
      data-testid="arrow"
    />
    <ul
      class="autocomplete-items"
      data-testid="items"
    >
      <li
        class="autocomplete-item"
        data-testid="li"
      >
        <div
          class="item-content"
        >
          <div
            class="avatar"
          >
            <div
              class="avatar-component"
            >
              <div
                class="avatar-alt"
                data-testid="alt"
              >
                HM
              </div>
              <img
                alt="Harriet McKinney"
                class="avatar-image hidden"
                data-testid="image"
              />
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="name"
            >
              Harriet McKinney
            </div>
            <div
              class="email"
            />
          </div>
        </div>
      </li>
      <li
        class="autocomplete-item"
        data-testid="li"
      >
        <div
          class="item-content"
        >
          <div
            class="avatar"
          >
            <div
              class="avatar-component"
            >
              <div
                class="avatar-alt"
                data-testid="alt"
              >
                MS
              </div>
              <img
                alt="Mathilda Summers"
                class="avatar-image hidden"
                data-testid="image"
              />
            </div>
          </div>
          <div
            class="info"
          >
            <div
              class="name"
            >
              Mathilda Summers
            </div>
            <div
              class="email"
            >
              alta.maxwell@kinetar.com
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
